n3fit/src/n3fit/vpinterface.py

```diff
@@ -331,3 +333,57 @@ def compute_arclength(self, q0=1.65, basis="evolution", flavours=None):
         flavours = ["sigma", "gluon", "V", "V3", "V8"]
     ret = arc_lengths(self, [q0], basis, flavours)
     return ret.stats.central_value()
+
+
+def compute_phi2(n3pdf, experimental_data):
+    """Compute phi2 using validphys functions.
+
+    For more info on how phi is calculated; see Eq.(4.6) of 10.1007/JHEP04(2015)040
+
+    Parameters
+    ----------
+        n3pdfs: :class:`n3fit.vpinterface.N3PDF`
+            `N3PDF` instance defining the n3fitted multi-replica PDF
+        experimental_data: List[validphys.core.DataGroupSpec]
+            List of experiment group datasets as `DataGroupSpec` instances
+
+    Returns
+    -------
+        sum_phi2: float
+            Sum of phi2 over all experimental group datasets
+
+    Example
+    -------
+    >>> from n3fit.vpinterface import N3PDF, compute_phi2
+    >>> from n3fit.model_gen import generate_pdf_model
+    >>> from validphys.loader import Loader
+    >>> fake_fl = [{'fl' : i, 'largex' : [0,1], 'smallx': [1,2]} for i in ['u', 'ubar', 'd', 'dbar', 'c', 'g', 's', 'sbar']]
+    >>> pdf_model = generate_pdf_model(nodes=[8], activations=['linear'], seed=0, num_replicas=2, flav_info=fake_fl, fitbasis="FLAVOUR")
+    >>> n3pdf = N3PDF(pdf_model.split_replicas())
+    >>> ds = Loader().check_dataset("NMC", theoryid=399, cuts="internal")
+    >>> data_group_spec = Loader().check_experiment("My DataGroupSpec", [ds])
+    >>> phi2 = compute_phi2(n3pdf, [data_group_spec])
+    """
+    sum_phi2 = 0.0
+    # Loop over the list of `DataGroupSpec` objects
+    for datagroupspec in experimental_data:
+        # datagroupspec is an instance of `DataGroupSpec`
+
+        # Loop over `DataGroupSpec` datasets
+        for datasetspec in datagroupspec.datasets:
+            # datasetspec is an instance of `DataSetSpec`
+
+            # get covariant matrix for each `DataSetSpec`
+            covmat = covmat_from_systematics(datasetspec.load_commondata(), datasetspec)
+
+            # get experiment (`DataResult`) and theory (`ThPredictionsResult`) predictions
+            res = results(datasetspec, n3pdf, covmat, sqrt_covmat(covmat))
+
+            # calculate standard chi2 (all_chi2) and chi2 using PDF central values (central_chi2)
+            chi2 = abs_chi2_data(res)
+
+            # calculate phi and store phi**2
+            phi, _ = phi_data(chi2)
+            sum_phi2 += phi**2
+
+    return sum_phi2
```

Comment:
```suggestion
            # calculate phi and store phi**2
            phi, ndat = phi_data(chi2)
            sum_phi += np.sqrt(ndat)*phi
            ndat_tot += ndat

    return sum_phi/np.sqrt(ndat_tot)
```
Where you initialise `ndat_tot=0` outside the loop. 

Since it's a secondary measure and only the hierarchy in phi value between different trials is what's used in hyperopt it doesn't really make a difference whether we use phi or phi2. However, looking at the paper where it says "an indicator which measures the standard deviation over the replica sample in units of the data uncertainty." we may want to go with phi.

Reply:
I am not sure if I understood well but I will accept anyway. Will this affect later on our measure of inverse phi2 in the model selection ? I am asking because this is the function that will calculate all phi2 per replica and fold as output in the `tries.json` file.

Reply:
I agree the comment was a bit confusing as I was talking about one thing but the change mainly addresses something I don't mention :sweat_smile:. The two things I change here are: 
1. compute phi instead of phi2 
2. change how phi2 is computed, see also this function: https://github.com/NNPDF/nnpdf/blob/162380804ca964da6de4c8ce47acddc36f25543b/validphys2/src/validphys/results.py#L678

The first one doesn't make much difference for the hyperopt we do since the absolute value of the phi measure is irrelevant and we just want to order trials by their value of phi (which is the same order if we were to use phi2). 

The second one actually changes the computation of phi.

Do you agree?

Reply:
Thanks @RoyStegeman. It it now clear from the docs in `total_phi_data_from_experiments`. Just changed then the name of `compute_phi2` to `compute_phi`

Reply:
Thanks!

Reply:
Done @RoyStegeman!